test/performance/tiledb_submit_a_b.cc

```diff
@@ -0,0 +1,305 @@
+  TRY(ctx, tiledb_query_alloc(ctx, array, TILEDB_READ, &b_query));
+  TRY(ctx, tiledb_query_set_layout(ctx, b_query, TILEDB_GLOBAL_ORDER));
+  TRY(ctx, tiledb_query_set_config(ctx, b_query, b_config));
+  json2query(ctx, array, b_query, query_config);
+
+  // helper to do basic checks on both
+  auto do_submit = [&](auto& key, auto& query, auto& outdata) -> uint64_t {
+    // make field size locations
+    auto dimension_sizes = []<typename... Ds>(std::tuple<Ds...> dimensions) {
+      return query_applicator<Asserter, Ds...>::make_field_sizes(dimensions);
+    }(outdata.dimensions());
+    auto attribute_sizes = []<typename... As>(std::tuple<As...> attributes) {
+      return query_applicator<Asserter, As...>::make_field_sizes(attributes);
+    }(outdata.attributes());
+
+    // add fields to query
+    [&]<typename... Ds>(std::tuple<Ds...> dims) {
+      query_applicator<Asserter, Ds...>::set(
+          ctx, query, dimension_sizes, dims, dimension_name);
+    }(outdata.dimensions());
+    [&]<typename... As>(std::tuple<As...> atts) {
+      query_applicator<Asserter, As...>::set(
+          ctx, query, attribute_sizes, atts, attribute_name);
+    }(outdata.attributes());
+
+    {
+      tiledb::sm::stats::DurationInstrument<TimeKeeper> qtimer =
+          time_keeper.start_timer(key);
+      TRY(ctx, tiledb_query_submit(ctx, query));
+    }
+
+    tiledb_query_status_t status;
+    TRY(ctx, tiledb_query_get_status(ctx, query, &status));
+
+    const uint64_t dim_num_cells = [&]<typename... Ds>(auto dims) {
+      return query_applicator<Asserter, Ds...>::num_cells(
+          dims, dimension_sizes);
+    }(outdata.dimensions());
+    const uint64_t att_num_cells = [&]<typename... As>(auto atts) {
+      return query_applicator<Asserter, As...>::num_cells(
+          atts, attribute_sizes);
+    }(outdata.attributes());
+
+    ASSERTER(dim_num_cells == att_num_cells);
+
+    if (dim_num_cells < outdata.size()) {
+      ASSERTER(status == TILEDB_COMPLETED);
+    }
+
+    return dim_num_cells;
+  };
+
+  const std::string a_key = std::string(array_uri) + ".a";
+  const std::string b_key = std::string(array_uri) + ".b";
+
+  while (true) {
+    const uint64_t a_num_cells = do_submit(a_key, a_query, a);
+    const uint64_t on_num_cells = do_submit(b_key, b_query, b);
+
+    ASSERTER(a_num_cells == on_num_cells);
+
+    std::apply(
+        [&](auto&... outfield) { (outfield.resize(a_num_cells), ...); },
+        std::tuple_cat(a.dimensions(), a.attributes()));
+    std::apply(
+        [&](auto&... outfield) { (outfield.resize(on_num_cells), ...); },
+        std::tuple_cat(b.dimensions(), b.attributes()));
+
+    ASSERTER(a.dimensions() == b.dimensions());
+
+    // NB: this is only correct if there are no duplicate coordinates,
+    // in which case we need to adapt what CSparseGlobalOrderFx::run does
+    ASSERTER(a.attributes() == b.attributes());
+
+    reset(a);
+    reset(b);
+
+    if (a_num_cells < num_user_cells) {
```

Comment:
what does this condition mean? that it's an `INCOMPLETE` query?

Reply:
That's correct.  In the unit tests I had written down that condition, I'm not sure why it didn't make it over here.

Actually I think there may be a bug here too, if this is never true then it will loop again and re-start the query from position zero (I think?)

It'll be better if I return the status from `do_submit` and we check it here.

Reply:
1a6f45fb